python/tvm/driver/tvmc/runner.py

```diff
@@ -112,8 +112,10 @@ def drive_run(args):
     except IOError as ex:
         raise TVMCException("Error loading inputs file: %s" % ex)
 
-    outputs, times = run_module(
-        args.FILE,
+    tvmc_package = TVMCPackage(package_path=args.FILE)
```

Comment:
I see this as a breaking change with the existing output format. I'm flagging this here to see what others think on whether we should maintain compatibility with our existing custom format or we might as well do the move and use the official APIs and abandon the custom implementation.

Reply:
You're right that this change means that packages created by tvmc before this PR will no longer be usable. After talking to @areusch, I'm of the understanding that the plan is to standardize model packaging using `export_model_library_format` across TVM soon, so I think this is a pain point that will have to be introduced into TVMC eventually. I thought it made sense to bite the bullet now rather than later. However, it would be pretty straight-forward to convert a legacy package to the new format. I'd be happy to provide a helper script that does that if you'd like.

Reply:
oh uhh, that's an idea I want to propose, but it hasn't been adopted by RFC or anything. apologies if I miscommunicated that. there are some things we need to work through there around BYOC I think before we can propose it, but I think we can probably accommodate a `.so` file in Model Library Format with a version bump.

Reply:
@areusch, I added .so exporting in this PR. Would you prefer i avoid using export_model_library_format for now?

Reply:
well I think this thread was originally about the format produced by `--output` (@leandron, correct me if I miss something here).

I think it would be to use the metadata and parameter export formats from model_library_format here. I see three areas of tension with doing this:
1. `codegen` is intended to be an exploded copy of the `.so`, so by adding the `.so`, it creates some confusion as to what exactly is expected to be here. it seems like we should be able to accommodate both source and binary forms in Model Library Format, but it seems like we need to think through things like: a) how should code consuming this format know whether it should expect a `.so` or source? b) what is the method by which you build the source in `codegen` into a binary? in having `codegen` source in Model Library Format, it would almost seem as if you should be able to build it with e.g. a `make` command. But if that's so, also having a binary pre-built seems weird without some additional explanation of what each thing is for, and unit tests that verify those stories.
2. doing this will expose Model Library Format to a lot more format revisions (which are different from the breaking change identified by @leandron here iiuc). Model Library Format revisions would impact downstream codegen using the microTVM Project API.
3. the one I mentioned, which is that Model Library Format doesn't support non-DSO-exportable Modules e.g. CUDA, Vulkan, etc.

i'm not sure what our policy is on maintaining `tvmc` output formats. I think adding more to the output of tvmc makes sense, but we may need to choose a command-line flag reflective of the format. I think also, perhaps given the tensions with Model Library Format, we should either
a) implement something a bit different here, so there is no confusion
b) do an RFC to resolve at least issue 1, perhaps we can come to consensus on a way to place the .so in Model Library Format

maybe since it seems like we should preface this PR with an RFC explaining the API, option b isn't so bad? wdyt?

Reply:
Ok, it seems like theres enough uncertainty around the use of model_library_format that I'm going to pull it out of this PR and go back to the custom packing. We can consider moving to a standard format at a later time once we work out some of these details.

Reply:
My goal is that the user doesn't need to acknowledge or understand the packing format at all. TVMCModel and TVMCPackage completely handle it.

Reply:
@leandron, I've reverted this PR to using the existing custom packaging format in TVMC. Existing packages should now be fully compatible after this refactor.

Reply:
Thanks @jwfromm I'll have a look.

(the comment got a bit long, but I thought to put the idea here and see what you think, as we probably got everybody in the PR anyway).

@areusch I think we have the opportunity to make Model Library Format (MLF) to be the way we serialize TVM outputs. Obviously, due to the nature of the outputs we produce from different backends (object files, shared libraries, sources), it can contain various artifacts within a tarball, and for that I think we could make better use of the metadata descriptor file, to describe what is in there in a way that we programmatically understand, for a given MLF package.

This would open the door to, for example, on Project API, we check whether a given tarball is created from a "micro" target and can be used to integrate with an RTOS. In TVMC, we can define whether the package aims, cpu, cuda, cl, etc. and whether it includes the shared library, etc.

With an abstraction like `TVMCPackage` (maybe renamed) we could make those features to be easily available for target applications already using TVM, to make use of the packages.

> the one I mentioned, which is that Model Library Format doesn't support non-DSO-exportable Modules e.g. CUDA, Vulkan, etc.

I think I saw you mentioning this limitation in the RFC as well, but I'm curious to understand a bit more on where is the limitation? is that on TVM side, or on the targets, such as Vulkan, that we can't export them?

Reply:
@leandron It's a bit of a long explanation, I will see if I can summarize. Currently a codegen backend is expected to produce runtime::Module. When either the load mechanism (if an existing e.g. CSourceModule could possibly be used) or the format (when producing e.g. not C) differs from one checked-in, a new Module must be defined. For this reason, many codegen produce aggregate modules containing multiple pieces of data. See CUDA for an example.

Worse yet, Modules are expected to implement their own load() and save(), and it is possible (though not likely now) that you could produce a Module in codegen which, after save() and load(), don't behave the same. It's possible we are not entirely unit testing now what we deploy.

Finally, Modules aren't labelled--so while it kind of makes sense how we've populated Model Library Format given target_host-targeted LLVM and C modules, it's difficult to implement any sort of generic organizational scheme.

My opinion is that the solution to this problem is generating a set of POD types (e.g. structs) from the compiler with labels and overhauling the module load process to better align it with the implied load process we are asking users to implement in microTVM. Specifically, it should be easy to document a set of steps that the user needs to implement on their own in order to arrive at an e.g. GraphExecutor from a given SoC after device reset.